### ISSUE TITLE
Possible fix for edge artifact in SpriteList batch draw

### ISSUE THREAD

**Issue by gcordonn** [pvcraven/arcade#660]
First, thanks for this nice package! I just discovered it - I used to work with heavier 3D engines, and I needed to do some 2D experiments, that's super useful. 

I quite quickly bumped into the edge artifact problem, and after looking around it seems like it already raised quite a lot of discussion, so please apologize if I am proposing something that was already rejected. 

It is a pity that was cannot use arbitrary transform or viewport resizing, especially with a package that is based on OpenGL, which is supposed to handle all of these smoothly. 

I quickly looked at the code, and it seems like the sprite list is using a texture atlas to store all textures used during batch rendering. The induced gain in batch-processing comes with issues with this technics, as discussed in http://download.nvidia.com/developer/NVTextureSuite/Atlas_Tools/Texture_Atlas_Whitepaper.pdf.

In particular, as already explained in Arcade documentation, the issue comes from the bilinear filtering: when a texel is sampled at a noninteger position, it gets interpolated with the neighboring one - which is somewhere in the texture atlas but sometimes out of the texture of the rendered sprite. 

One (easy) fix is to add a .5 offset to the texture coordinate (before normalization), in the _calculate_sub_tex_coords function of SpriteList.py, so that the texel samples are located at the middle of the pixel. 

```python
            offset=.5

            for index, image in enumerate(self.array_of_images):
                column = index % grid_width
                row = index // grid_width

                # Texture coordinates are reversed in y axis
                row = grid_height - row - 1

                x = column * (grid_item_width + MARGIN) +offset
                y = row * (grid_item_height + MARGIN) +offset

                # Because, coordinates are reversed
                y += (grid_item_height - (image.height - MARGIN))

                normalized_x = x / sprite_sheet_width
                normalized_y = y / sprite_sheet_height

                start_x = normalized_x
                start_y = normalized_y

                normalized_width = (image.width-2*offset) / sprite_sheet_width
                normalized_height = (image.height-2*offset) / sprite_sheet_height
```

But this solution will give results that are not consistent with the current behavior of arcade (although I think this is the more 'correct' for perfect pixel alignment.)

Then there is a second solution, that would also be preferred should mipmapping be supported, which is to pad the textures in the sprite map with a one-pixel border:

```python
                if name_of_texture_to_check not in new_array_of_texture_names:
                    new_array_of_texture_names.append(name_of_texture_to_check)
                    if sprite.texture is None:
                        raise ValueError(f"Sprite has no texture.")
                    if sprite.texture.image is None:
                        raise ValueError(f"Sprite texture {sprite.texture.name} has no image.")
                    image = sprite.texture.image

                    ########## FIX HERE ############
                    tmp = Image.new('RGBA', (image.width+2, image.height+2))
                    tmp.paste(image, (1,1))
                    tmp.paste(tmp.crop((1          , 1           , image.width+1, 2             )), (1            , 0             ))
                    tmp.paste(tmp.crop((1          , image.height, image.width+1, image.height+1)), (1            , image.height+1))
                    tmp.paste(tmp.crop((1          , 0           ,             2, image.height+2)), (0            , 0             ))
                    tmp.paste(tmp.crop((image.width, 0           , image.width+1, image.height+2)), (image.width+1, 0             ))
```

and set the previously added offset to 1:

```
# ...
offset=1
# ...
```
This needs two more pixels in w and y per image in the texture atlas. the atlas already adds a margin to each image `MARGIN=1` that can be removed with these two methods.  The later solution gives results which are exactly similar that the current behavior of SpriteList on 'good' rendering conditions, but suppress the edges artifacts when they used to appear.  

I didn't test that extensively so there might be some issues remaining. 

I am not opening a pull request because I see there is some ongoing work on OpenGL improvements - and maybe this is colliding?

**Comment by einarf** [comment]
I think this is worth trying out, but ultimately that is up to Paul. You could PR to `shader_experimental` branch because this is were all the action is currently.

> It is a pity that was cannot use arbitrary transform or viewport resizing, especially with a package that is based on OpenGL, which is supposed to handle all of these smoothly.

You can hack around this by setting `arcade.window_commands._projection` manually or using `arcade.create_orthogonal_projection(..)` and set viewport with (pyglet) `gl.glViewport`. Right now the `set_viewport` actually sets the projection and sets the viewport to the buffer size. It's on my list of things to figure out. Probably adding a "camera" that would separate between game resolution and buffer resolution. supporting zooming, back borders making resizing a no-brainer for most.

**Comment by pvcraven** [comment]
The 2.4 shader_experimental branch would be the place to try this out. The edge artifacts are a serious headache to programmers, that I wish didn't have to worry about it.

**Comment by gcordonn** [comment]
Ok, I am quite busy these days, but let me try that. Is there any backward compatibility requirement?

In particular, if the viewport/transform is set to have an exact 1 to 1 correspondence with the size of the window, do you want the pixels of the sprite to fall exactly on the pixel of the window? 
That's a behavior I was not able to obtain even after disabling the AA, (not sure that I don't have any additional Nvidia or windows postprocessing antialiasing).

**Comment by einarf** [comment]
If you are extremely busy I can experiment with uv offsets to achieve this. I'm going over the sprite list soon anyway and possibly also looking at viewport/projection stuff, so this would be a good combo.

**Comment by pvcraven** [comment]
Released in 2.4.
